docs/concepts/time-and-windows.rst

```diff
@@ -0,0 +1,122 @@
+.. _time-and-windows:
+
+Time and Windows
+################
+
+During the life cycle of a record, it experiences a number of events

```

Comment:
I think this is a little confusing.  The record is the event, it doesn't experience events.  Nor does it have a lifecycle in this context.

Maybe just have:

> When talking about time it is important to be clear about _which_ time is being discussed. The time an event occurred may be different to the time it was ingested into Kafka, or the time at which the processing of the event is occurring. 

???

Reply:
I'd like to get very crisp about this -- my thinking is that in the KSQL world, "events" and "messages" are abstracted away, and we have instead "records". But maybe this isn't the way we want to think about it...

Reply:
I agree, that there is nothing like a life cycle.

I also agree, that we should abstract over events/messages and use the term records -- we use records in KS docs, too, for the same reason. "Message" is the standard term for core Kafka, ie, pub/sub.

Reply:
I'm fine with 'records'. The reason I mentioned 'events' is because of the use of 'events' here:

> it experiences a number of events

The record does not experience anything.  It is an immutable bit of data / information. 

From a KSQL point of view, a record has a timestamp, (i.e. the timestamp of the record in Kafka). This is the default timestamp that KSQL will use.  This may be an 'event-time' or just the 'ingestion-time' depending on what the record's producer did when producing the record to Kafka. (The producer can chose to set the timestamp to any time they like, but most often chose either some sensible 'event-time' or the current time, which would be equivalent to 'ingestion-time'. If they don't set a timestamp its set to 'ingestion-time' by Kafka.)

When importing the topic in to KSQL it will, by default, using this timestamp for the record. Alternatively, a user can use the `WITH(TIMESTAMP='some-field')` to tell KSQL to use an alternative timestamp from the record's value, (with an optional `TIMESTAMP_FORMAT`).  The field the user specifies may be an 'event-time' or a 'ingestion-time'.

I think I'm right in saying KSQL currently doesn't support any 'processing-time' operations, (though they're probably possible with UDFs to access the current time).

So.. from KSQL perspective, I don't think the _type_ of the timestamp, i.e. event vs ingestion vs processing time, really matters that much. KSQL just uses the timestamp that you tell it too.

That said, a little background explaining that different concepts of time exist and that its important to know which one you're discussing is probably worthwhile as a preamble, just don't talk about records having a lifecycle or experiencing events ;)

Reply:
> This may be an 'event-time' or just the 'ingestion-time' depending on what the record's producer did when producing the record to Kafka. (The producer can chose to set the timestamp to any time they like, but most often chose either some sensible 'event-time' or the current time, which would be equivalent to 'ingestion-time'. If they don't set a timestamp its set to 'ingestion-time' by Kafka.)

IMHO, this is not 100% correct (with respect to out current terminology). (1) brokers allow to configure topic with either `CreateTime` (default) vs `LogAppendTime` -- for the first config, the timestamp will be whatever the producer sets. For the second config, the broker *always* set the timestamp. The first case implies "ingestion time" semantics. Thus, if LogAppendTime is configures, the producer has no control whatsoever over the timestamp.

If `CreateTime` is set, the following holds for the producer:
 - if a `ProducerRecord` is created, by default it contains no timestamp
 - users can set the timestamp on `ProducerRecord` explicitly though
 - if the timestamp is not set, on producer.send() the current wall-clock time will be set automatically

Note, that we consider all three cases as "event-time" and don't distinguish them in our current terminology. I agree, that semantically we could follow Andy's suggestion, but I *believe* this would make it more complex for users to understand the subtle differences. Thus, I would prefer to stay with our current (maybe simplified) terminology.

>  For example, things will go awry if some records were written to kafka with the field containing some event-time, which others go the default ingest-time set by the broker.

I guess this is not possible.

However, another important thing to point out: if a `WITH(timestamp=...)` clause is used, this timestamp must be passed to Kafka Streams as "unix epoch time in milli seconds" -- not sure how much control the user has about this, and what part KSQL handles. If the use could mess this up (ie, KSQL cannot ensure this timestamp format/semantic) we need to tell users about it.

Reply:
This is all great stuff, please keep it coming! We'll get a KSQL-focused topic out of this yet.